### PR TITLE
Add jextract option to disable buildID check

### LINKED ISSUE #10676: Optional workaround for jextract RAS build id difference when processing core file
There was a crash on plinux in the builds last night (https://github.com/eclipse/openj9/issues/10588#issuecomment-697467680), something we've been chasing. I want to jextract the core file and get the system libraries so we can look at it (the crashing machine doesn't have gdb installed). The builds last night don't contain the latest jextract fix https://github.com/eclipse/openj9/pull/10661, when I tried to use a build containing the fix, I got the following error.
```
java.lang.Error: J9RAS.buildID is incorrect (found bca9b0bb00000039, expecting bca9b0bb000003a5). This version of jextract is incompatible with this dump.
	at com.ibm.jvm.j9.dump.extract.Main.getEnvironmentPointer(Native Method)
	at com.ibm.jvm.j9.dump.extract.Main.<init>(Main.java:469)
	at com.ibm.jvm.j9.dump.extract.Main.main(Main.java:342)
Error. Unable to locate executable for core.20200923.055935.1753.0001.dmp
```

I think there is no reason a different buildID can't be used to process the core file and collect the system libraries. Discussing with Keith on slack, he suggested this is advanced behavior, as it could end up collecting the wrong set of shared libraries for the jvm if the original jvm has been replaced at the same location. 

We can still have a workaround to disable the check, in the form of a system properly like `jextract -J-Djextract.disable.J9RAS.buildID.check <core>` to be used by advanced users only. Keith suggested printing a warning is appropriate when using this option, which I agree with.

### PR DESCRIPTION
**Pull request by keithc-ca**
Disable check if `jextract.disable.J9RAS.buildID.check` is set (to any value).

* always query environment pointer before analyzing dump
* clean up `jextract` support code
* generate `J9VM_UNIQUE_BUILD_ID` randomly for cmake builds

Fixes #10676.

### CODE REVIEW

**Comment by pshipton** [comment]
Please create a docs issue and PR to document the option.

File: runtime/jextractnatives/jextractnatives.c [hunk 1]
```diff
@@ -461,37 +455,38 @@ flushCache(void)
...
+	static UDATA total;
 
 	dbgCacheElement* thisElement = NULL;
 	UDATA lineStart = address & ~(UDATA)(sizeof(thisElement->data) - 1);
 	UDATA endAddress = address + size;
 
 #if 0
-	if ( ((++total) % (16 * 1024)) == 0) {
-		dbgPrint("Cache hit rate: %.2f\n", (float)hits / (float)(total));
+	if (((++total) % (16 * 1024)) == 0) {
```
Comment (pshipton) [inline runtime/jextractnatives/jextractnatives.c:472]: This is using uninitialized variables. Also `total` is unused outside of this disabled block of code.

**Comment by pshipton** [comment]
@manqingl are you ok with the jextract change https://github.com/eclipse/openj9/issues/10676#issuecomment-702160809

File: runtime/jextractnatives/jextractnatives.c [hunk 2]
```diff
@@ -127,70 +120,76 @@ validateDump(JNIEnv *env)
...
 					(*env)->ThrowNew(env, errorClazz, errBuf);
 					return JNI_FALSE;
 				}
 				if (ras->buildID != J9UniqueBuildID) {
-					j9str_printf(PORTLIB, 
-						errBuf, sizeof(errBuf), 
-						"J9RAS.buildID is incorrect (found %llx, expecting %llx). This version of jextract is incompatible with this dump.", 
+					j9str_printf(PORTLIB,
+						errBuf, sizeof(errBuf),
+						"J9RAS.buildID is incorrect (found %llx, expecting %llx). This version of jextract is incompatible with this dump.",
```
Comment (pshipton) [inline runtime/jextractnatives/jextractnatives.c:181]: I think the option to disable the check should be mentioned in this message.

**Comment by keithc-ca** [comment]
> Please create a docs issue and PR to document the option.

See https://github.com/eclipse/openj9-docs/issues/649

File: runtime/jextractnatives/jextractnatives.c (hunk 2 above)
Comment (keithc-ca) [reply runtime/jextractnatives/jextractnatives.c:181]: I thought of that, but I wasn't sure how obscure the bypass should be. If we're going to explicitly call it out, perhaps it should just be a normal command-line option?

File: runtime/jextractnatives/jextractnatives.c (hunk 1 above)
Comment (keithc-ca) [reply runtime/jextractnatives/jextractnatives.c:472]: The C language guarantees that static variables are initialized to zero.
I'll make that explicit and make all the statistical stuff conditional.

File: runtime/jextractnatives/jextractnatives.c (hunk 2 above)
Comment (pshipton) [reply runtime/jextractnatives/jextractnatives.c:181]: I'm not sure how big the impact to service will be. If it turns out some customers are running jextract from a different JVM version, we should mitigate the impact by making it easy to find the workaround. I don't have strong feelings either way for sticking with the current option or making it a command line option.

Comment (keithc-ca) [reply runtime/jextractnatives/jextractnatives.c:181]: It's both easier to talk about and use if it's a command-line option: how about `-r` (relax checking)?
I'll also document the remaining supported options.

Comment (pshipton) [reply runtime/jextractnatives/jextractnatives.c:181]: Seems consistent with the other options in the help, -f and -v.

**Comment by keithc-ca** [comment]
A preview of changes in progress:
```
$ jextract core.20201001.130835.2036.0001.dmp
Loading dump file...
java.lang.Error: J9RAS.buildID is incorrect (found 7036b61000000000, expecting db377fa63ff6cb25). This version of jextract is incompatible with this dump (use '-r' option to relax this check).
	at openj9.dtfj/com.ibm.jvm.j9.dump.extract.Main.getEnvironmentPointer(Native Method)
	at openj9.dtfj/com.ibm.jvm.j9.dump.extract.Main.<init>(Main.java:442)
	at openj9.dtfj/com.ibm.jvm.j9.dump.extract.Main.main(Main.java:328)
Error. Unable to locate executable for core.20201001.130835.2036.0001.dmp
```
and with `-r`:
```
$ jextract -r core.20201001.130835.2036.0001.dmp
Loading dump file...
Ignoring incorrect J9RAS.buildID (found 7036b61000000000, expecting db377fa63ff6cb25). This version of jextract may be incompatible with this dump.
VM set to 00007F92E800E880
Read memory image from core.20201001.130835.2036.0001.dmp
Creating archive file: core.20201001.130835.2036.0001.dmp.zip
Adding "core.20201001.130835.2036.0001.dmp"
[...]
```

**Comment by keithc-ca** [comment]
Updated based on review feedback.

File: jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java [hunk 3]
```diff
@@ -367,20 +353,18 @@ private static void ensure(boolean condition, String errorMessage) {
...
-			}
-			System.exit(code);
-		}
+		report("Usage: jextract [options] dump_name [output_filename]"); //$NON-NLS-1$
+		report("  output filename defaults to dump_name.zip"); //$NON-NLS-1$
+		report("  options:"); //$NON-NLS-1$
+		report("    -e                  throw exceptions instead of calling System.exit()"); //$NON-NLS-1$
+		report("    -f executable_name  override executable name"); //$NON-NLS-1$
+		report("    -help               print this usage message"); //$NON-NLS-1$
+		report("    -interactive        start an interactive session"); //$NON-NLS-1$
```
Comment (pshipton) [inline jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java:362]: This is no longer useful, it was replaced by DDR. We kept it alive after DDR was available to support the trprint subcommand, but trying to run that, it doesn't seem to work very well any more and I don't think anybody uses it. Bottom line, it shouldn't be included in the help, and I expect we could delete all the associated code.

File: jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java [hunk 4]
```diff
@@ -367,20 +353,18 @@ private static void ensure(boolean condition, String errorMessage) {
...
+		report("    -p prefix           prefix for all paths (absolute or relative)"); //$NON-NLS-1$
+		report("    -r                  relax version checking"); //$NON-NLS-1$
+		report("    -v                  enable verbose output"); //$NON-NLS-1$
+		report("    --                  mark end of options"); //$NON-NLS-1$
```
Comment (pshipton) [inline jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java:366]: I don't understand the purpose of this, perhaps a more detailed message is appropriate.

File: jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java (hunk 3 above)
Comment (keithc-ca) [reply jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java:362]: Ok, I won't mention `-interactive`. I created #10764 to track removing the option and the support infrastructure.

File: jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java (hunk 4 above)
Comment (keithc-ca) [reply jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java:366]: This is a commonly supported option with the meaning that subsequent arguments are not treated as options (even if they begin with `-`). A (perhaps unlikely) example:
```
jextract -r -- -v
```
considers the file named `-v` to be the core file.

Comment (keithc-ca) [reply jcl/src/openj9.dtfj/share/classes/com/ibm/jvm/j9/dump/extract/Main.java:366]: I think it's a GNU convention; see the [bash manual](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Invoking-Bash), for example.

**Comment by keithc-ca** [comment]
Removed mention of '-interactive' option in help.

**Comment by pshipton** [comment]
Running internal testing.

**Comment by manqingl** [comment]
@pshipton @keithc-ca : jextract not only gets the OS libraries, it also gets the ones from Java itself.  Will this change get the wrong Java libraries?

**Comment by pshipton** [comment]
The VM_Extended.testDTFJ jobs are failing, see build id 456590.

**Comment by pshipton** [comment]
@manqingl it won't get the wrong libraries, it may cause an error like the following, where no error occurred before. The user would need to re-run with the `-r` option.
`java.lang.Error: J9RAS.buildID is incorrect (found 2e855c200006f78e, expecting 7036b6100006f78e). This version of jextract is incompatible with this dump (use '-r' option to relax this check).`

**Comment by manqingl** [comment]
> 
> 
> @manqingl it won't get the wrong libraries, it may cause an error like the following, where no error occurred before. The user would need to re-run with the `-r` option.
> `java.lang.Error: J9RAS.buildID is incorrect (found 2e855c200006f78e, expecting 7036b6100006f78e). This version of jextract is incompatible with this dump (use '-r' option to relax this check).`

@pshipton : thanks for the clarification.  I have no other concerns.

**Comment by keithc-ca** [comment]
I think the test failures are because the build ids are different for compressed and non-compressed references.
I'll test that theory in an internal build.

**Comment by pshipton** [comment]
Seems always checking the buildID should wait until we have a single JVM (no separate compressed and XL builds).

**Comment by keithc-ca** [comment]
See internal build 456637.

**Comment by keithc-ca** [comment]
Update with new definition of UMA's `unique_build_id`.

**Comment by pshipton** [comment]
I'm not sure of the side affects of making compressedrefs and XL ids the same. We'll have to look at all the places which use the build ids and consider what the side affects might be.

**Comment by keithc-ca** [comment]
`J9UniqueBuildID` is only used in two places:
 1) in [rasdump.c](https://github.com/eclipse/openj9/blob/0606174bd898b28531f89460c57411072c3b29da/runtime/vm/rasdump.c#L321) so it is captured in core files
 2) in [jextractnatives.c](https://github.com/eclipse/openj9/blob/2f295ea91cdd97ff37eb5d164fc63014ea718bc7/runtime/jextractnatives/jextractnatives.c#L186) to verify that it matches the value read from a core file

**Comment by pshipton** [comment]
jenkins test sanity,extended win,zlinux jdk8

**Review by pshipton — APPROVED** [review]

**Pull request merged**